stdlib/http/src/main/java/org/ballerinalang/net/http/compiler/ResourceValidator.java

```diff
@@ -2,68 +2,172 @@
+                    break;
+                case ANN_NAME_QUERY_PARAM:
+                    validateQueryParam(param, dlog);
+                    break;
+                case ANN_NAME_BODY_PARAM:
+                    if (bodyParamCount++ == 0) {
+                        validateBodyParam(param, dlog);
+                        continue;
+                    }
+                    dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "invalid multiple `" + HTTP_ANNOTATION +
+                            ANN_NAME_BODY_PARAM + "` annotations: cannot specify > 1 entity-body params");
+                    break;
+                default:
+                    // do not execute
+            }
+        }
+    }
+
+    private static boolean isInvalidResourceParam(BLangSimpleVariable param, String expectedType) {
+        return !expectedType.equals(param.type.toString());
+    }
+
+    private static String getCompatibleAnnotation(BLangSimpleVariable param) {
+        for (BLangAnnotationAttachment annotationAttachment : param.annAttachments) {
+            BAnnotationSymbol annotationSymbol = annotationAttachment.annotationSymbol;
+            if (!PROTOCOL_HTTP.equals(annotationSymbol.pkgID.name.value)) {
+                continue;
+            }
+            String annotationName = annotationSymbol.name.value;
+            if (ANN_NAME_PATH_PARAM.equals(annotationName) || ANN_NAME_QUERY_PARAM.equals(annotationName) ||
+                    ANN_NAME_BODY_PARAM.equals(annotationName)) {
+                return annotationName;
+            }
+        }
+        return null;
+    }
+
+    private static void validatePathParam(BLangSimpleVariable param, List<String> pathSegments, DiagnosticLog dlog) {
+        BType paramType = param.type;
+        int varTag = paramType.tag;
+
+        if (varTag != TypeTags.STRING && varTag != TypeTags.INT && varTag != TypeTags.BOOLEAN &&
+                varTag != TypeTags.FLOAT) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible path param type: `" + param.type +
+                    "`, expected `string`, `int`, `boolean`, `float`");
+            return;
+        }
+        if (!pathSegments.contains(param.name.value)) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "invalid path param: `" + param.toString() +
+                    "`, missing segment `{" + param.name.value + "}` in the path config of the resource annotation");
+        }
+    }
+
+    private static void validateQueryParam(BLangSimpleVariable param, DiagnosticLog dlog) {
+        BType paramType = param.type;
+        int varTag = paramType.tag;
+
+        if (varTag != TypeTags.STRING && (varTag != TypeTags.ARRAY ||
+                ((BArrayType) paramType).getElementType().tag != TypeTags.STRING)) {
+            dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible query param type: `" + param.type +
+                    "`, expected `string`, `string[]`");
```

Comment:
Shall we change this to the same format as the standard incompatible types error message? 
```java
dlog.logDiagnostic(Diagnostic.Kind.ERROR, param.pos, "incompatible query param type: expected 'string|string[]', found '" + param.type + "'";
```
Also, the expected type can be stated as `string|string[]` right? Also, how would the user know what type(s) to use for query params?
Another point, in error messages, we use ' around type names, not `

Reply:
Changed the error as suggested.
If we stated the type as `string|string[]`, user can misinterpret is as a union type as ballerina has similar syntax. 
The types of each param are mentioned in the module.md. Also, we can add it to the annotation docs if suitable

Reply:
Yeah assuming we only allow either `string` or `string[]` I also think `string|string[]` could be misleading.

How about
`"incompatible query param: expected a param of type 'string' or 'string[]', found '" + param.type + "'"`

On  a separate note, does "query **parameter**" sound odd? `QueryParam` for the annotation is OK, but I was wondering if we should use "parameter" for the error message.

Reply:
Changed the error as suggested